src/api/js/src/high-level/high-level.ts

```diff
@@ -0,0 +1,1636 @@
+    return `${major}.${minor}.${build_number}`;
+  }
+
+  function getFullVersion() {
+    return Z3.get_full_version();
+  }
+
+  function openLog(filename: string) {
+    return Z3.open_log(filename);
+  }
+
+  function appendLog(s: string) {
+    Z3.append_log(s);
+  }
+
+  function setParam(key: string, value: any): void;
+  function setParam(params: Record<string, any>): void;
+  function setParam(key: string | Record<string, any>, value?: any) {
+    if (typeof key === 'string') {
+      Z3.global_param_set(key, value.toString());
+    } else {
+      assert(value === undefined, "Can't provide a Record and second parameter to set_param at the same time");
+      Object.entries(key).forEach(([key, value]) => setParam(key, value));
+    }
+  }
+
+  function resetParams() {
+    Z3.global_param_reset_all();
+  }
+
+  function getParam(name: string) {
+    return Z3.global_param_get(name);
+  }
+
+  function isContext(obj: unknown): obj is Context {
+    return obj instanceof ContextImpl;
+  }
+
+  class ContextImpl {
+    declare readonly __typename: Context['__typename'];
+
+    readonly ptr: Z3_context;
+    readonly name: string;
+
+    constructor(name: string, params: Record<string, any> = {}) {
+      const cfg = Z3.mk_config();
+      Object.entries(params).forEach(([key, value]) => Z3.set_param_value(cfg, key, value.toString()));
+      const context = Z3.mk_context_rc(cfg);
+
+      this.ptr = context;
+      this.name = name;
+
+      Z3.set_ast_print_mode(this.ptr, Z3_ast_print_mode.Z3_PRINT_SMTLIB2_COMPLIANT);
+      Z3.del_config(cfg);
+
+      this.AstRef = AstRefImpl.bind(AstRefImpl, this) as any;
+      this.Solver = SolverImpl.bind(SolverImpl, this);
+      this.Model = ModelImpl.bind(ModelImpl, this);
+      this.FuncInterp = FuncInterpImpl.bind(FuncInterpImpl, this);
+      this.SortRef = SortRefImpl.bind(SortRefImpl, this);
+      this.FuncDeclRef = FuncDeclRefImpl.bind(FuncDeclRefImpl, this);
+      this.ExprRef = ExprRefImpl.bind(ExprRefImpl, this) as any;
+      this.BoolSortRef = BoolSortRefImpl.bind(BoolSortRefImpl, this);
+      this.BoolRef = BoolRefImpl.bind(BoolSortRefImpl, this);
+      this.PatternRef = PatternRefImpl.bind(PatternRefImpl, this);
+      this.Probe = ProbeImpl.bind(ProbeImpl, this);
+      this.Tactic = TacticImpl.bind(TacticImpl, this);
+      this.ArithSortRef = ArithSortRefImpl.bind(ArithSortRefImpl, this);
+      this.ArithRef = ArithRefImpl.bind(ArithRefImpl, this);
+      this.IntNumRef = IntNumRefImpl.bind(IntNumRefImpl, this);
+      this.RatNumRef = RatNumRefImpl.bind(RatNumRefImpl, this);
+      this.AstVector = AstVectorImpl.bind(AstVectorImpl, this) as any;
+      this.AstMap = AstMapImpl.bind(AstMapImpl, this) as any;
+
+      // We want to bind functions and operations to `this` inside Context
+      // So that the user can write things like this and have it work:
+      // ```
+      // const { And, Or } = new Context('main');
```

Comment:
If you make everything closed over rather than accessing properties of `this`, then you don't need to `bind` anything and this code will still work. That will be a _much_ nicer implementation.

Reply:
Could you explain more on what you mean?

The API should still work like this
```typescript
const ctx = new Context('main');
```

Reply:
`bind` is only necessary for methods which refer to `this`, and there is no reason to do so for this API.

That is, if instead of having

```js
class ContextImpl {
  _assertContext(...ctxs: (Context | { ctx: Context })[]) {
    ctxs.forEach(other => assert('ctx' in other ? this === other.ctx : this === other, 'Context mismatch'));
  }

  isAst(obj: unknown): obj is Ast {
    const r = obj instanceof AstImpl;
    r && this._assertContext(obj);
    return r;
  }
}
return { Context: ContextImpl };
```
you had
```js
function createContext(name) {
  let ctx = new ContextImpl(name);
  function _assertContext(...ctxs: (Context | { ctx: Context })[]) {
    ctxs.forEach(other => assert('ctx' in other ? ctx === other.ctx : ctx === other, 'Context mismatch'));
  }
  return {
    isAst(obj: unknown): obj is Ast {
      const r = obj instanceof AstImpl;
      r && _assertContext(obj); // note lack of `this`
      return r;
    }
  };
}
return { Context: createContext };
```
then everything would work without `bind`. And, in fact, you could still write `const ctx = new Context('main')`, if you really wanted to.